autograd.h

```diff
@@ -297,22 +297,50 @@ AUTOGRAD_CONTAINER_CLASS(Conv) {
 
 class Conv1d : public Conv {
  public:
-  Conv1d(uint32_t i, uint32_t o, int ks) : Conv(1, i, o, ks) { } 
+  Conv1d(uint32_t i, uint32_t o, int ks) : Conv(1, i, o, ks) { }
   Conv1d(uint32_t i, uint32_t o, IntVec ks) : Conv(1, i, o, ks) { }
 };
 
 class Conv2d : public Conv {
  public:
-  Conv2d(uint32_t i, uint32_t o, int ks) : Conv(2, i, o, ks) { } 
+  Conv2d(uint32_t i, uint32_t o, int ks) : Conv(2, i, o, ks) { }
   Conv2d(uint32_t i, uint32_t o, IntVec ks) : Conv(2, i, o, ks) { }
 };
 
 class Conv3d : public Conv {
  public:
-  Conv3d(uint32_t i, uint32_t o, int ks) : Conv(3, i, o, ks) { } 
+  Conv3d(uint32_t i, uint32_t o, int ks) : Conv(3, i, o, ks) { }
   Conv3d(uint32_t i, uint32_t o, IntVec ks) : Conv(3, i, o, ks) { }
 };
 
+AUTOGRAD_CONTAINER_CLASS(BatchNorm) {
+ public:
+  BatchNorm(uint32_t num_features)
+    : num_features_(num_features) {}
+
+  AUTOGRAD_KWARG(BatchNorm, double, eps, 1e-5, 1e-5)
+  AUTOGRAD_KWARG(BatchNorm, double, momentum, 0.1, 0.1)
+  AUTOGRAD_KWARG(BatchNorm, bool, affine, true, true)
+  AUTOGRAD_KWARG(BatchNorm, bool, stateful, false, true)
+
+  void reset_parameters() override;
+  variable_list forward(variable_list) override;
+  void initialize_parameters() override;
+
+  Variable weight;
+  Variable bias;
+  Variable running_mean;
+  Variable running_var;
+
+ protected:
+  uint32_t num_features_;
+};
+
+class BatchNorm1d : public BatchNorm {
```

Comment:
Not much point in implementing BatchNorm1d and 2d actually, if they're treated the same.